### ISSUE TITLE
Add `calc` support to `declaration-property-value-no-unknown`

### ISSUE THREAD

**Issue by silverwind** [stylelint/stylelint#8891]
### What is the problem you're trying to solve?

[`declaration-property-value-no-unknown`](https://stylelint.io/user-guide/rules/declaration-property-value-no-unknown/) is lacking `calc` support. This is detected:

```css
a { height: 2 }
```

but this is not:

```css
a { height: calc(2) }
```

### What solution would you like to see?

Determine the resulting data type of `calc` and then check if it is valid for the property it is being assigned on. Some examples:

```css
a { height: calc(2) } /* ❌ */
a { height: calc(2 + 2) } /* ❌ */
a { height: calc(2 * 2) } /* ❌ */
a { height: calc(2px + 2) } /* ❌ */
a { height: calc(2px * 2) }  /* ✅ */
a { height: calc(var(--foo) + 2) } /* ❌ (assuming variable is unknown) */
a { height: calc(var(--foo) * 2) } /* ✅ (assuming variable is unknown) */
```

**Comment by romainmenke** [comment]
Hi @silverwind,

Thank you for opening this issue.

As far as I can tell `csstree` (which we use in this rule) doesn't actually validate the contents of a `calc()` function.

These are all seemingly valid for `csstree`:

[Demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcICGACYAdAdugLgCxgEsBzPHeFMJAGzAAoBGFAahQCYAHADwEoBuLLgIkyFKrToAzCBH6D8RUuUo16zdermYAviAA0ISYWowAckgC2cRDC6WOJgHRgAzi-3gImI8QQhbHBAATjgosJJIAK7UoRjYOEHRMC4UcbjpILBg1EhBSDiEXgC0HEEQHDAhAJ5FAG40kTBFmBBFkZgA1i0A7tgIKAmN8tpY2gKYHpDeJABiwRb5fi44VSbUhJg4zl4+jhYAVu4GsBzuiHEoKOggy6sw65vXFNc5OMk413pYl9e3axs4IpTHxFZZITBQXJQJ5XECvd7XUYgbRAA)

```css
a {
	height: calc(1 + 2px);
	height: calc(foo);
	height: calc(+++++);
}
```

See: https://github.com/csstree/csstree/blob/56afb6dd761149099cd3cdfb0a38e15e8cc0a71a/lib/lexer/generic.js#L144-L147

```js
// TODO: implement
// can be used wherever <length>, <frequency>, <angle>, <time>, <percentage>, <number>, or <integer> values are allowed
// https://drafts.csswg.org/css-values/#calc-notation
function calc(next) {
```

-------

There is also the hurdle that the CSS specification and syntax definitions aren't sufficiently detailed or precise.

For example keywords that are only valid in specific properties or functions (e.g. `r`, `g`, `b` in relative `rgb()`) can't be expressed in syntax definitions. Following what can be expressed in syntax definitions these would either be valid in all math functions, everywhere, or in none of them.

[So fixing this completely also requires a new way to define such constraints in the CSS specifications.](https://github.com/w3c/csswg-drafts/issues/11505)

-------

I did create a package to solve math expressions: https://github.com/csstools/postcss-plugins/tree/main/packages/css-calc#css-calc-

In theory this can be used as a way to validate syntax.
i.e. If an expression can't be solved, it isn't valid.

Keywords can also be defined:

```js
import { calc } from '@csstools/css-calc';

const globals = new Map([
	['a', '10px'],
	['b', '2rem'],
]);

// '20px'
console.log(calc('calc(a * 2)', { globals: globals }));
```

So it kinda does some of the work that is needed to resolve this issue.

But it has the mirror problem as `csstree`, it can only validate the contents of a math expression. It can't verify that the result of that math expression is actually valid where it is being used.

So it can detect `calc(1px + 1)` but it can't report `rgb(calc(1px + 2px) 0 0)`

Given how complex this was to implement I do understand why this is missing in `csstree` 😅

**Comment by silverwind** [comment]
So this rule is validated in `csstree` only? Wouldn't it be better to process the AST in stylelint code? Is the AST even available to stylelint?

**Comment by romainmenke** [comment]
`csstree` also has all the data of what is valid and what isn't and all the logic to apply that data to an AST.

Re-implementing that would kind of defeat the purpose of using `csstree` :)

But the AST is available and custom validations are possible.
It is mostly that improving validation upstream in `csstree` makes more sense because that has a broader impact.
